src/System.Management.Automation/security/Authenticode.cs

```diff
@@ -578,70 +591,81 @@ private static X509Certificate2 GetCertFromChain(IntPtr pSigner)
             DWORD error,
             NativeMethods.WINTRUST_DATA wtd)
         {
+            s_tracer.WriteLine("GetSignatureFromWintrustData: error: {0}", error);
+
             Signature signature = null;
-            X509Certificate2 signerCert = null;
+            IntPtr pProvSigner = IntPtr.Zero;
             X509Certificate2 timestamperCert = null;
+            if (TryGetProviderSigner(wtd.hWVTStateData, ref pProvSigner, ref timestamperCert))
+            {
+                //
+                // get cert of the signer
+                //
+                X509Certificate2 signerCert = GetCertFromChain(pProvSigner);
 
-            s_tracer.WriteLine("GetSignatureFromWintrustData: error: {0}", error);
+                if (signerCert != null)
+                {
+                    if (timestamperCert != null)
+                    {
+                        signature = new Signature(filePath,
+                                                  error,
+                                                  signerCert,
+                                                  timestamperCert);
+                    }
+                    else
+                    {
+                        signature = new Signature(filePath,
+                                                  error,
+                                                  signerCert);
+                    }
+
+                    signature.SignatureType = SignatureType.Authenticode;
+                }
+            }
+
+            Diagnostics.Assert(((error == 0) && (signature != null)) || (error != 0), "GetSignatureFromWintrustData: general crypto failure");
+
+            if ((signature == null) && (error != 0))
+            {
+                signature = new Signature(filePath, error);
+            }
+
+            return signature;
+        }
 
+        [ArchitectureSensitive]
+        private static bool TryGetProviderSigner(IntPtr wvtStateData, ref IntPtr pProvSigner, ref X509Certificate2 timestamperCert)
```

Comment:
It seems to me both `pProvSigner` and `timestamperCert` are better to be an out parameter instead of a ref parameter.